src/JSInterop/Microsoft.JSInterop/src/DotNetDispatcher.cs

```diff
@@ -276,16 +334,48 @@ private static (MethodInfo, Type[]) GetCachedMethodInfo(string assemblyName, str
             return result;
         }
 
-        private static Assembly GetRequiredLoadedAssembly(string assemblyName)
+        private static Assembly GetRequiredLoadedAssembly(AssemblyKey assemblyKey)
         {
             // We don't want to load assemblies on demand here, because we don't necessarily trust
             // "assemblyName" to be something the developer intended to load. So only pick from the
             // set of already-loaded assemblies.
             // In some edge cases this might force developers to explicitly call something on the
             // target assembly (from .NET) before they can invoke its allowed methods from JS.
             var loadedAssemblies = AppDomain.CurrentDomain.GetAssemblies();
-            return loadedAssemblies.FirstOrDefault(a => a.GetName().Name.Equals(assemblyName, StringComparison.Ordinal))
-                ?? throw new ArgumentException($"There is no loaded assembly with the name '{assemblyName}'.");
+            return loadedAssemblies.FirstOrDefault(a => new AssemblyKey(a).Equals(assemblyKey))
+                ?? throw new ArgumentException($"There is no loaded assembly with the name '{assemblyKey.AssemblyName}'.");
         }
+
+        private readonly struct AssemblyKey : IEquatable<AssemblyKey>
```

Comment:
I'm working around an issue with the arcade \ xunit: https://github.com/dotnet/arcade/issues/2816. Essentially the test has two instances of JSInterop.Tests in the AppDomain. The first one is only used by xunit and the latter by the test code.  

This results in weird errors:

> [A]TestDTO cannot be cast to [B]TestDTO. Type A originates from 'Microsoft.JSInterop.Tests, Version=42.42.42.42, Culture=neutral, PublicKeyToken=adb9793829ddae60' in the context 'Default' at location '/Users/vsts/agent/2.150.3/work/1/s/artifacts/bin/Microsoft.JSInterop.Tests/Debug/netcoreapp3.0/Microsoft.JSInterop.Tests.dll'. Type B originates from 'Microsoft.JSInterop.Tests, Version=42.42.42.42, Culture=neutral, PublicKeyToken=adb9793829ddae60' in the context 'Default' at location '/Users/vsts/agent/2.150.3/work/1/s/artifacts/bin/Microsoft.JSInterop.Tests/Debug/netcoreapp3.0/Microsoft.JSInterop.Tests.dll'.

However it seems generally OK to match by Assembly identity when it's available. If you'd like I can file an issue to clean this up later or just leave this as-is

Reply:
Ah, you're trying to solve *that* problem.